lib/burnside/flattened.h

```diff
@@ -0,0 +1,264 @@
+struct SwitchOp
+  : public SumTypeCopyMixin<const parser::CallStmt *,
+        const parser::ComputedGotoStmt *, const parser::ArithmeticIfStmt *,
+        const parser::CaseConstruct *, const parser::SelectRankConstruct *,
+        const parser::SelectTypeConstruct *> {
+  template<typename A>
+  SwitchOp(
+      const A &sw, const std::vector<LabelRef> &refs, const Location &source)
+    : SumTypeCopyMixin{&sw}, refs{refs}, source{source} {}
+  void dump() const;
+
+  const std::vector<LabelRef> refs;
+  Location source;
+};
+
+// a compute step
+struct ActionOp {
+  ActionOp(const parser::Statement<parser::ActionStmt> &stmt) : v{&stmt} {}
+  void dump() const;
+
+  const parser::Statement<parser::ActionStmt> *v;
+};
+
+#define CONSTRUCT_TYPES \
+  const parser::AssociateConstruct *, const parser::BlockConstruct *, \
+      const parser::CaseConstruct *, const parser::ChangeTeamConstruct *, \
+      const parser::CriticalConstruct *, const parser::DoConstruct *, \
+      const parser::IfConstruct *, const parser::SelectRankConstruct *, \
+      const parser::SelectTypeConstruct *, const parser::WhereConstruct *, \
+      const parser::ForallConstruct *, const parser::CompilerDirective *, \
+    const parser::OpenMPConstruct * , const parser::OmpEndLoopDirective *
+
+// entry into a Fortran construct
+struct BeginOp : public SumTypeCopyMixin<CONSTRUCT_TYPES> {
+  SUM_TYPE_COPY_MIXIN(BeginOp)
+  void dump() const;
+
+  template<typename A> BeginOp(const A &c) : SumTypeCopyMixin{&c} {}
+};
+
+// exit from a Fortran construct
+struct EndOp : public SumTypeCopyMixin<CONSTRUCT_TYPES> {
+  SUM_TYPE_COPY_MIXIN(EndOp)
+  void dump() const;
+
+  template<typename A> EndOp(const A &c) : SumTypeCopyMixin{&c} {}
+};
+
+struct DoIncrementOp {
+  DoIncrementOp(const parser::DoConstruct &stmt) : v{&stmt} {}
+  void dump() const;
+
+  const parser::DoConstruct *v;
+};
+
+struct DoCompareOp {
+  DoCompareOp(const parser::DoConstruct &stmt) : v{&stmt} {}
+  void dump() const;
+
+  const parser::DoConstruct *v;
+};
+
+// the flat structure is a list of Ops, where an Op is any of ...
+struct Op : public SumTypeMixin<LabelOp, GotoOp, ReturnOp, ConditionalGotoOp,
+                SwitchIOOp, SwitchOp, ActionOp, BeginOp, EndOp, IndirectGotoOp,
+                DoIncrementOp, DoCompareOp> {
+  template<typename A> Op(const A &thing) : SumTypeMixin{thing} {}
+
+  void dump() const {
+    std::visit([](const auto &op) { op.dump(); }, u);
+  }
+
+  static void Build(std::list<Op> &ops,
+      const parser::Statement<parser::ActionStmt> &ec, AnalysisData &ad);
+};
+
+// helper to build unique labels
+struct LabelBuilder {
+  LabelBuilder();
+  LabelRef getNext();
+  void setReferenced(LabelRef label);
+  bool isReferenced(LabelRef label) const;
+  std::vector<bool> referenced;
+  unsigned counter;
+};
+
+LabelOp FetchLabel(AnalysisData &ad, const parser::Label &label);
+
+std::vector<LabelRef> GetAssign(
+    AnalysisData &ad, const semantics::Symbol *symbol);
+}
```

Comment:
A comment on this brace marking it as the end of `namespace flat` would be nice.

Reply:
ok